### PR TITLE
Generic Microsoft Graph API Wrapper

### LINKED ISSUE #329: [Extensibility Request / Idea] Graph API Module
### BC Idea Link

https://experience.dynamics.com/ideas/idea/?ideaid=2972d343-f481-ee11-a81c-6045bd860893

### Description

**🕮 Describe the issue**
I created this issue to discuss about an improvement for the system application by providing an Graph API module for communction with the Microsoft Graph API.

As a starter there should be some kind of IMsGraphAuthorization interface.

And a RestWrapper (MsGraphAPIClient to Post/Patch/Delete request to any specified endpoint of the graph API.

Maybe some Authoriziation Codeunits can be reused in this module.
I'm thinking about the BlobStorage Authorization or the SharePoint Authorization.

I would be willing to provide a first draft for this module, but I could need some help on writing tests, since I don't know to test generic external API calls.

**Additional context**
Original PR on ALAppExtensions: https://github.com/microsoft/ALAppExtensions/pull/22919 <br /><br />Internal work item: AB#477759

### PR DESCRIPTION
**Pull request by pri-kise**
This is a first draft for a generic microsoft graph api wrapper as mentioned in issue https://github.com/microsoft/BCApps/issues/329

This PR replaces the PR (https://github.com/microsoft/ALAppExtensions/pull/22919) on the old repository ALAppExtensions.

Fixes #329 

Work item: AB#477759

### CODE REVIEW

File: src/System Application/App/MicrosoftGraph/app.json
Comment (mazhelez) [inline src/System Application/App/MicrosoftGraph/app.json]: Versions around here should probably be 24

File: src/System Application/App/Rest Client/src/Authentication/HttpAuthOAuthClientCredentials.Codeunit.al
Comment (jwikman) [inline src/System Application/App/Rest Client/src/Authentication/HttpAuthOAuthClientCredentials.Codeunit.al]: Where is `AuthCodeErrGlobal` assigned a value?

File: src/System Application/App/MicrosoftGraph/app.json
Comment (JesperSchulz) [reply src/System Application/App/MicrosoftGraph/app.json]: I've fixed that.

File: src/System Application/App/Rest Client/src/Authentication/HttpAuthOAuthClientCredentials.Codeunit.al
Comment (pri-kise) [reply src/System Application/App/Rest Client/src/Authentication/HttpAuthOAuthClientCredentials.Codeunit.al]: Updated this and removed this variable

**Comment by JesperSchulz** [comment]
Security Review scheduled for Tuesday, 12th of December.

**Comment by JesperSchulz** [comment]
Security review conducted with @darjoo. Darrick will comment on outcome.

**Comment by darjoo** [comment]
After performing the security review with Jesper, the following is the outcome that needs to be fixed.

1. Telemetry for all usages
Fx, telemetry when calling `Get`, `Post`...
OnCall, OnSuccess, OnError

2. Removal of `SetBaseUrl`.
As far as I can tell, Graph doesn't use custom domains (Do correct me if I'm wrong). Ensuring that we're always calling graph's url, keeps our customer's access and data safe.

**Comment by JesperSchulz** [comment]
@pri-kise, if you can address the feedback from the security review next week, I can make sure your module releases with 23.3!
On a different note: I'm not super happy about the mg abbreviation. Any chance we could change that to e.g. Graph? It's not a must, it just hurts my eyes 😋

**Comment by pri-kise** [comment]
> After performing the security review with Jesper, the following is the outcome that needs to be fixed.
> 
> 1. Telemetry for all usages
>    Fx, telemetry when calling `Get`, `Post`...
>    OnCall, OnSuccess, OnError
@darjoo Could you please help me on this topic?
Where do you expect to be an telemtry event?

Do you want some for the RestClient Module?

or do you want one at each process method in `codeunit 9351 "Mg Client Impl."`?
```al
procedure Get(RelativeUriToResource: Text; MgOptionalParameters: Codeunit "Mg Optional Parameters"; var HttpResponseMessage: Codeunit "Http Response Message"): Boolean
var
    MicrosoftGraphUriBuilder: Codeunit "Mg Uri Builder";
    FeatureTelemetry: Codeunit "Feature Telemetry";
    MicrosoftGraphCategoryLbl: Label 'Connect to Azure Functions', Locked = true;
begin
    Clear(HttpResponseMessage);
    FeatureTelemetry.LogUsage('0000Mg000', MicrosoftGraphCategoryLbl, 'TDB'); //Note: Details needed
    MicrosoftGraphUriBuilder.Initialize(MicrosoftGraphBaseUrl, MicrosoftGraphAPIVersion, RelativeUriToResource, MgOptionalParameters.GetQueryParameters());
    MgRequestHelper.SetRestClient(RestClient);
    HttpResponseMessage := MgRequestHelper.Get(MicrosoftGraphUriBuilder, MgOptionalParameters);
    exit(HttpResponseMessage.GetIsSuccessStatusCode());
end;
```
And do you maybe have an example of another module where this telemetry pattern is applied currently?
I've checked some and got very different patterns for FeatureTelemetry.
What you expect for the EventId and for the Event Category?
And what information am I allowed to add as custom dimensions, that don't impact security?
I could really need a short call to clarify this request.

> 2. Removal of `SetBaseUrl`.
>    As far as I can tell, Graph doesn't use custom domains (Do correct me if I'm wrong). Ensuring that we're always calling graph's url, keeps our customer's access and data safe.

As far as I can tell, there are scenarios where a custom national domain could be used to access microsoft graph.
https://learn.microsoft.com/en-us/graph/deployments

@JesperSchulz I took the _Mg_ abbreviation from the PowerShell Module for Microsoft Graph. I neede a short abbreviation and this one is used at least one other Microsoft Tool.

**Comment by pri-kise** [comment]
I applied the requested Changes

**Comment by darjoo** [comment]
Here are the discussion points from our call.

- Telemetry is no longer required to be added to the module as we get sufficient telemetry out of the HttpClient itself. Which in-turn also means the Rest Client does not need to put in addition telemetry.
- Keep the SetBaseUrl as well. This can be tracked from HttpClient's outgoing hostname and there is valid reason for other base urls, even though it's strictly the a handful.

**Comment by JesperSchulz** [comment]
That means we are now on the final stretch! I'll run CI. Assuming it passes, we'll get 2 code review sign-offs, and then we'll hit the merge button!
I'll then create a PR for 23.3 and we'll get this brand new module out with the January release 🥳

**Review by JesperSchulz — APPROVED** [review]

**Review by darjoo — APPROVED** [review]

**Comment by devgar** [comment]
Hi, nice job, I'm working to implement this codeunit in our code.

Doing that we found that there's no PUT mehtod implemented in this wrapper. Is it scheluded for later? We require it to perform sharepoint content updates.

**Comment by JesperSchulz** [comment]
No further additions are planned. Feel free to create an issue describing your requirements. Once approved, you are free to implement the required changes yourself!

**Pull request merged**
